pkg/controllers/dynakube/extension/eec/reconciler.go

```diff
@@ -70,3 +84,106 @@ func (r *reconciler) Reconcile(ctx context.Context) error {
 
 	return r.createOrUpdateStatefulset(ctx)
 }
+
+func (r *reconciler) reconcileTLSSecret(ctx context.Context) error {
+	query := k8ssecret.Query(r.client, r.client, log)
+
+	secret, err := query.Get(ctx, types.NamespacedName{
+		Name:      getSelfSignedTLSSecretName(r.dk.Name),
+		Namespace: r.dk.Namespace,
+	})
+
+	if err != nil && !k8serrors.IsNotFound(err) {
+		return err
+	}
+
+	if r.dk.Spec.Templates.ExtensionExecutionController.TlsRefName != "" {
+		if secret != nil {
+			return query.Delete(ctx, secret)
+		}
+
+		return nil
+	}
+
+	if k8serrors.IsNotFound(err) {
+		err = r.createOrUpdateTLSSecret(ctx)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	if secret != nil {
+		err = r.reconcileTLSSecretExpiration(ctx, secret)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *reconciler) createOrUpdateTLSSecret(ctx context.Context) error {
+	cert, err := certificates.New()
+	if err != nil {
+		return err
+	}
+
+	cert.Cert.DNSNames = getCertificateAltNames(r.dk.Name)
+	cert.Cert.KeyUsage = x509.KeyUsageKeyEncipherment | x509.KeyUsageDataEncipherment
+	cert.Cert.ExtKeyUsage = []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth}
+	cert.Cert.Subject.CommonName = r.dk.Name + "-extensions-controller.dynatrace"
+
+	err = cert.SelfSign()
+	if err != nil {
+		return err
+	}
+
+	pemCert, pemPk, err := cert.ToPEM()
+	if err != nil {
+		return err
+	}
+
+	coreLabels := k8slabels.NewCoreLabels(r.dk.Name, k8slabels.ExtensionComponentLabel)
+	secretData := map[string][]byte{consts.TLSCrtDataName: pemCert, consts.TLSKeyDataName: pemPk}
+
+	secret, err := k8ssecret.Build(r.dk, getSelfSignedTLSSecretName(r.dk.Name), secretData, k8ssecret.SetLabels(coreLabels.BuildLabels()))
+	if err != nil {
+		return err
+	}
+
+	secret.Type = corev1.SecretTypeTLS
+
+	query := k8ssecret.Query(r.client, r.client, log)
+
+	_, err = query.CreateOrUpdate(ctx, secret)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *reconciler) reconcileTLSSecretExpiration(ctx context.Context, secret *corev1.Secret) error {
+	isValid, err := certificates.ValidateCertificateExpiration(secret.Data[consts.TLSCrtDataName], consts.ExtensionsSelfSignedTLSRenewalThreshold, time.Now(), log)
```

Comment:
can you please use timeProvider like other reconcilers, instead of hardcoding time.Now(), it will help to unit-test it in future.

Reply:
Fixed